apps/small_filters/process.h

```diff
@@ -0,0 +1,526 @@
+#endif
+
+#ifdef CONV3X3A32
+#include "conv3x3a32_hvx128.h"
+#include "conv3x3a32_hvx64.h"
+#include "conv3x3a32_cpu.h"
+#else
+DECL_WEAK_PIPELINE3S(conv3x3a32)
+#endif
+
+
+enum bmark_run_mode_t {
+    hvx64 = 1,
+    hvx128 = 2,
+    cpu = 3
+};
+
+template <typename T>
+T clamp(T val, T min, T max) {
+    if (val < min)
+        return min;
+    if (val > max)
+        return max;
+    return val;
+}
+
+typedef int (*pipeline2)(buffer_t *, buffer_t *);
+typedef int (*pipeline3)(buffer_t *, buffer_t *, buffer_t *);
+
+struct PipelineDescriptorBase {
+    virtual void init() = 0;
+    virtual const char * name() = 0;
+    virtual int run(bmark_run_mode_t mode) = 0;
+    virtual bool verify(int W, int H) = 0;
+    virtual bool defined() = 0;
+};
+
+template <typename T1>
+struct PipelineDescriptor : public PipelineDescriptorBase  {
+    T1 pipeline_64, pipeline_128, pipeline_cpu;
+ 
+    PipelineDescriptor(T1 pipeline_64, T1 pipeline_128, T1 pipeline_cpu) :
+    pipeline_64(pipeline_64), pipeline_128(pipeline_128), pipeline_cpu(pipeline_cpu) {}
+ 
+    void init() {
+        return;
+    }
+    const char *name() {
+        return "";
+    }
+    int run(bmark_run_mode_t mode) {
+        return -1;
+    }
+    bool verify(int W, int H) {
+        return false;
+    }
+    bool defined() {
+        return pipeline_64 && pipeline_128 && pipeline_cpu;
+    }
+};
+
+class Conv3x3a16Descriptor : public PipelineDescriptor<pipeline3> {
+    Halide::Runtime::Buffer<uint8_t> u8_in, u8_out;
+    Halide::Runtime::Buffer<int8_t> i8_mask;
+
+public:
+    Conv3x3a16Descriptor(pipeline3 pipeline_64, pipeline3 pipeline_128, pipeline3 pipeline_cpu,
+                         int W, int H) :
+        PipelineDescriptor<pipeline3>(pipeline_64, pipeline_128, pipeline_cpu),
+        u8_in(nullptr, W, H, 2),
+        u8_out(nullptr, W, H, 2),
+        i8_mask(nullptr, 3, 3, 2) {}
+
+    void init() {
+        u8_in.device_malloc(halide_hexagon_device_interface());
+        u8_out.device_malloc(halide_hexagon_device_interface());
+        i8_mask.device_malloc(halide_hexagon_device_interface());
+
+        u8_in.for_each_value([&](uint8_t &x) {
+            x = static_cast<uint8_t>(rand());
+        });
+        u8_out.for_each_value([&](uint8_t &x) {
+            x = 0;
+        });
+
+        i8_mask(0, 0) = 1;
+        i8_mask(1, 0) = -4;
+        i8_mask(2, 0) = 7;
+
+        i8_mask(0, 0) = 2;
+        i8_mask(1, 0) = -5;
+        i8_mask(2, 0) = 8;
+
+        i8_mask(0, 0) = 3;
+        i8_mask(1, 0) = -6;
+        i8_mask(2, 0) = 7;
+    }
+
+    const char *name() { return "conv3x3a16"; }
+
+    bool verify(const int W, const int H) {
+        u8_out.for_each_element([&](int x, int y) {
+            int16_t sum = 0;
+            for (int ry = -1; ry <= 1; ry++) {
+                for (int rx = -1; rx <= 1; rx++) {
+                    int clamped_x = (x + rx < 0) ? 0 : x + rx;
+                    clamped_x = (clamped_x >= W) ? (W-1) : clamped_x;
```

Comment:
Consider using your clamp function here, it would be easier to read:

    int clamped_x = clamp(x + rx, 0, W - 1);

Ditto the other descriptors.